s3driver.go

```diff
@@ -0,0 +1,216 @@
+package sftp
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+type S3 interface {
+	ListObjectsV2(input *s3.ListObjectsV2Input) (*s3.ListObjectsV2Output, error)
+	DeleteObject(input *s3.DeleteObjectInput) (*s3.DeleteObjectOutput, error)
+	CopyObject(input *s3.CopyObjectInput) (*s3.CopyObjectOutput, error)
+	PutObject(input *s3.PutObjectInput) (*s3.PutObjectOutput, error)
+	GetObject(input *s3.GetObjectInput) (*s3.GetObjectOutput, error)
+}
+
+type S3Driver struct {
+	s3       S3
+	bucket   string
+	homePath string
+}
+
+func (d S3Driver) Stat(path string) (os.FileInfo, error) {
+	localPath, err := translatePath(d.homePath, path)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := d.s3.ListObjectsV2(&s3.ListObjectsV2Input{
+		Bucket:  aws.String(d.bucket),
+		Prefix:  aws.String(localPath),
+		MaxKeys: aws.Int64(1),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.Contents == nil || *resp.KeyCount == 0 {
+		return nil, os.ErrNotExist
+	}
+
+	info := &fileInfo{
+		name: localPath,
+		mode: os.ModePerm,
+	}
+	if strings.HasSuffix(*resp.Contents[0].Key, "/") {
+		info.name = strings.TrimRight(info.name, "/")
+		info.mode = os.ModeDir
+	}
+	return info, nil
+}
+
+func (d S3Driver) ListDir(path string) ([]os.FileInfo, error) {
+	prefix, err := translatePath(d.homePath, path)
+	if err != nil {
+		return nil, err
+	}
+	objects, err := d.s3.ListObjectsV2(&s3.ListObjectsV2Input{
+		Bucket:    aws.String(d.bucket),
+		Prefix:    aws.String(prefix),
+		Delimiter: aws.String("/"),
+	})
+	if err != nil {
+		return nil, err
+	}
+	files := []os.FileInfo{}
+	for _, o := range objects.Contents {
+		if *o.Key == prefix {
+			continue
+		}
+		files = append(files, &fileInfo{
+			name:  strings.TrimPrefix(*o.Key, prefix),
+			size:  *o.Size,
+			mtime: *o.LastModified,
+		})
+	}
+	for _, o := range objects.CommonPrefixes {
+		files = append(files, &fileInfo{
+			name: strings.TrimSuffix(strings.TrimPrefix(*o.Prefix, prefix), "/"),
+			mode: os.ModeDir,
+		})
+	}
+	return files, nil
+}
+
+func (d S3Driver) DeleteDir(path string) error {
+	translatedPath, err := translatePath(d.homePath, path)
+	if err != nil {
+		return err
+	}
+	_, err = d.s3.DeleteObject(&s3.DeleteObjectInput{
+		Bucket: aws.String(d.bucket),
+		Key:    aws.String(translatedPath),
+	})
```

Comment:
I'm pretty sure a directory delete on S3 is not this simple http://stackoverflow.com/a/11427712